### PR TITLE
integer_arithmetic: detect integer arithmetic on references.

### LINKED ISSUE #5328: Clippy misses integer arithmetic when one operand is a reference
I expected this code to trigger an "integer arithmetic" warning:
```rust
#![warn(clippy::integer_arithmetic)]

pub fn foo(x: &i32) -> i32 {
    let a = 5;
    a+x
}
```
However, [no such warning is emitted](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=d93225ee13584f436b8bf8a26492691d).

### PR DESCRIPTION
**Pull request by matthiaskrgr**
changelog: integer_arithmetic fix false negatives with references on integers

Fixes #5328

### CODE REVIEW

File: clippy_lints/src/arithmetic.rs [hunk 1]
```diff
@@ -82,7 +82,8 @@ impl<'a, 'tcx> LateLintPass<'a, 'tcx> for Arithmetic {
                     _ => (),
                 }
                 let (l_ty, r_ty) = (cx.tables.expr_ty(l), cx.tables.expr_ty(r));
-                if l_ty.is_integral() && r_ty.is_integral() {
+                // warn on integers as well as references to integers
+                if l_ty.is_integral() || l_ty.is_any_ptr() && r_ty.is_integral() || r_ty.is_any_ptr() {
```
Comment (flip1995) [inline clippy_lints/src/arithmetic.rs]: Doesn't this now lint also for `&f32 + &f32`. Reading the [`is_any_ptr`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc/ty/struct.TyS.html#method.is_any_ptr) documentation:

> Tests if this is any kind of primitive pointer type (reference, raw pointer, fn pointer).

I think it does.

Comment (matthiaskrgr) [reply clippy_lints/src/arithmetic.rs]: Mmh good point. I believe I have a more elegant solution now.

File: tests/ui/arithmetic.stderr [hunk 2]
```diff
@@ -1,127 +1,211 @@
...
+LL |     f1 + f2
+   |     ^^^^^^^
+
+error: floating-point arithmetic detected
+  --> $DIR/arithmetic.rs:145:5
+   |
+LL |     (&f1 + &f2)
+   |     ^^^^^^^^^^^
+
+error: aborting due to 34 previous errors
```
Comment (flip1995) [inline tests/ui/arithmetic.stderr]: Wait this should fail the `--limit-stderr-length` test 🤔

Comment (flip1995) [reply tests/ui/arithmetic.stderr]: I have a fix for this.

Comment (matthiaskrgr) [reply tests/ui/arithmetic.stderr]: Should I split up the into two, one for float and one for int arithmetic?

Comment (flip1995) [reply tests/ui/arithmetic.stderr]: Would be good. I just pushed the fix for this here. So please pull before splitting up the tests.

**Comment by flip1995** [comment]
@bors r+

Thanks!

**Pull request merged**